File: 06-inherit/main.py
class SchoolMember:
    '''Base class (superclass)'''
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def tell(self):
        print(f'Name:"{self.name}" Age:"{self.age}"', end=" ")

class Teacher(SchoolMember):
    '''Subclass (derived class)'''
    def __init__(self, name, age, salary):
        SchoolMember.__init__(self, name, age)
        self.salary = salary

    def tell(self):
        SchoolMember.tell(self)
        print(f'Salary: "{self.salary}"')

class Student(SchoolMember):
    '''Subclass (derived class)'''
    def __init__(self, name, age, marks):          
        SchoolMember.__init__(self, name, age)
        self.marks = marks

    def tell(self):
        SchoolMember.tell(self)
        print(f'Marks: "{self.marks}"')

m = SchoolMember('Miklos', 20)
t = Teacher('Bond', 40, 30000)
s = Student('Moneypenny', 25, 75)

members = [m, t, s]
for member in members:
   member.tell()
   if member.name == 'Miklos':
       print()

File: 06-inherit/main.go
package main

import "fmt"

type Teller interface {
	tell()
	whoami() string
}

type SchoolMember struct {
	name string
	age  int
}

func (m SchoolMember) tell() {
	fmt.Printf("Name:\"%s\" Age:\"%d\"", m.name, m.age)
}

func (m SchoolMember) whoami() string {
	return m.name
}

type Teacher struct {
	sm     SchoolMember
	salary int
}

func (m Teacher) tell() {
	m.sm.tell()
	fmt.Printf(" Salary:\"%d\"\n", m.salary)
}

func (m Teacher) whoami() string {
	return m.sm.name
}

type Student struct {
	sm    SchoolMember
	marks int
}

func (m Student) tell() {
	m.sm.tell()
	fmt.Printf(" Marks:\"%d\"\n", m.marks)
}

func (m Student) whoami() string {
	return m.sm.name
}

func main() {
	m := SchoolMember{"Miklos", 20}
	t := Teacher{SchoolMember{"Bond", 40}, 30000}
	s := Student{SchoolMember{"Moneypenny", 25}, 75}
	members := []Teller{m, t, s}

	for _, member := range members {
		member.tell()
		if member.whoami() == "Miklos" {
			fmt.Println()
		}
	}
}
